File: server/otegami-relay/Sources/OtegamiRelay/Store/CredentialCrypto.swift
import Crypto
import Foundation

/// Encrypts/decrypts IMAP credentials at rest (`RelayStore`'s `watch` rows)
/// with AES-256-GCM, keyed by `RELAY_MASTER_KEY` (plan §7: "サーバ側は
/// RELAY_MASTER_KEY(env) の AES-GCM で暗号化保存"). The key never touches
/// disk itself — it's supplied only via the environment, so a stolen
/// database file alone is useless without also having compromised the
/// process's environment.
///
/// Device secrets are handled separately (`RelayStore` stores only their
/// SHA-256 hash, like a password) — this type is only ever used for IMAP
/// credentials, which the relay genuinely needs to recover in plaintext to
/// open the watch's IMAP connection.
struct CredentialCrypto: Sendable {
    enum CredentialCryptoError: Error, CustomStringConvertible {
        case invalidKeyLength(Int)
        case invalidBase64
        case decryptionFailed

        var description: String {
            switch self {
            case .invalidKeyLength(let length):
                "RELAY_MASTER_KEY must decode to exactly 32 bytes (got \(length))"
            case .invalidBase64:
                "RELAY_MASTER_KEY is not valid base64"
            case .decryptionFailed:
                "failed to decrypt stored credential (wrong key, or corrupted/tampered data)"
            }
        }
    }

    // Stored as raw `Data` rather than `SymmetricKey` directly: on Linux,
    // swift-crypto's `SymmetricKey` doesn't conform to `Sendable` (unlike
    // on Apple platforms, where `Crypto` re-exports CryptoKit's Sendable-
    // safe type) — this server only ever runs on Linux in production
    // (docs/relay-deployment.md), so a `SymmetricKey` stored property would
    // fail to build there even though it builds fine locally on macOS.
    // `SymmetricKey(data:)` is cheap to reconstruct per call.
    private let keyData: Data

    init(key: SymmetricKey) {
        self.keyData = key.withUnsafeBytes { Data($0) }
    }

    /// Parses the `RELAY_MASTER_KEY` env var: base64 for exactly 32 raw
    /// bytes (AES-256).
    init(base64Key: String) throws {
        guard let data = Data(base64Encoded: base64Key) else {
            throw CredentialCryptoError.invalidBase64
        }
        guard data.count == 32 else {
            throw CredentialCryptoError.invalidKeyLength(data.count)
        }
        self.keyData = data
    }

    private var key: SymmetricKey { SymmetricKey(data: keyData) }

    /// Generates a fresh random 32-byte key, base64-encoded — what
    /// `docs/relay-deployment.md` tells operators to run once
    /// (`openssl rand -base64 32`-equivalent) to produce `RELAY_MASTER_KEY`.
    static func generateBase64Key() -> String {
        SymmetricKey(size: .bits256).withUnsafeBytes { Data($0) }.base64EncodedString()
    }

    /// Encrypts `plaintext`, returning the AES-GCM sealed box's combined
    /// representation (nonce + ciphertext + tag) as a single `Data` — the
    /// only thing `RelayStore` needs to persist per credential.
    func encrypt(_ plaintext: String) throws -> Data {
        let sealed = try AES.GCM.seal(Data(plaintext.utf8), using: key)
        guard let combined = sealed.combined else {
            // Only possible if the nonce were non-standard-size, which
            // `AES.GCM.seal` never produces without an explicit custom
            // nonce being passed in (we don't).
            throw CredentialCryptoError.decryptionFailed
        }
        return combined
    }

    /// Reverses `encrypt(_:)`. Throws `.decryptionFailed` on any
    /// authentication-tag mismatch (wrong key or tampered/corrupted data) —
    /// deliberately not more specific, to avoid oracle-style error
    /// messages leaking anything about *why* it failed.
    func decrypt(_ combined: Data) throws -> String {
        do {
            let sealed = try AES.GCM.SealedBox(combined: combined)
            let plaintext = try AES.GCM.open(sealed, using: key)
            guard let string = String(data: plaintext, encoding: .utf8) else {
                throw CredentialCryptoError.decryptionFailed
            }
            return string
        } catch is CredentialCryptoError {
            throw CredentialCryptoError.decryptionFailed
        } catch {
            throw CredentialCryptoError.decryptionFailed
        }
    }
}

File: server/otegami-relay-go/internal/cryptox/credential_crypto.go
// Package cryptox encrypts/decrypts IMAP credentials at rest, exactly
// compatible with the retired Swift relay's CredentialCrypto.
//
// Compatibility is the single most delicate part of this port (Task #180):
// an existing SQLite database has `watch.encryptedSecret` BLOB columns
// written by swift-crypto's `AES.GCM.seal(_:using:)`, and this relay must
// decrypt them with the same RELAY_MASTER_KEY without any migration step.
//
// swift-crypto's `AES.GCM.SealedBox.combined` representation (what
// CredentialCrypto.encrypt stores) is, per Apple's CryptoKit/swift-crypto
// documentation: nonce (12 bytes) || ciphertext (same length as plaintext)
// || authentication tag (16 bytes), concatenated as raw bytes — no
// base64, no length prefixes, no additional authenticated data.
//
// Go's crypto/cipher.AEAD (via crypto/aes + cipher.NewGCM, which defaults
// to a 12-byte nonce and a 16-byte tag, matching AES-GCM's standard sizes)
// produces the identical layout when Seal's `dst` argument is the nonce
// itself: `aead.Seal(nonce, nonce, plaintext, nil)` appends
// ciphertext||tag onto `dst`, yielding nonce||ciphertext||tag — byte-for-
// byte what swift-crypto's `.combined` produces for the same key/nonce/
// plaintext. This is verified against real Swift-encrypted fixtures in
// credential_crypto_test.go (TestDecryptsSwiftEncryptedFixture and
// TestDecryptsSwiftEncryptedUTF8Fixture).
package cryptox

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"unicode/utf8"
)

// nonceSize and tagSize match AES-GCM's standard sizes, which is what both
// swift-crypto's AES.GCM (no custom nonce ever passed by CredentialCrypto)
// and Go's cipher.NewGCM (default constructor) use.
const (
	nonceSize = 12
	tagSize   = 16
)

// ErrInvalidKeyLength mirrors CredentialCryptoError.invalidKeyLength.
type ErrInvalidKeyLength struct{ Length int }

func (e ErrInvalidKeyLength) Error() string {
	return fmt.Sprintf("RELAY_MASTER_KEY must decode to exactly 32 bytes (got %d)", e.Length)
}

// ErrInvalidBase64 mirrors CredentialCryptoError.invalidBase64.
var ErrInvalidBase64 = errors.New("RELAY_MASTER_KEY is not valid base64")

// ErrDecryptionFailed mirrors CredentialCryptoError.decryptionFailed —
// deliberately generic (wrong key vs. tampered/corrupted data are not
// distinguished), matching the Swift side's oracle-avoidance rationale.
var ErrDecryptionFailed = errors.New("failed to decrypt stored credential (wrong key, or corrupted/tampered data)")

// CredentialCrypto encrypts/decrypts IMAP credentials with AES-256-GCM,
// keyed by RELAY_MASTER_KEY.
type CredentialCrypto struct {
	key []byte // exactly 32 bytes
}

// New wraps an already-decoded 32-byte AES-256 key.
func New(key []byte) (*CredentialCrypto, error) {
	if len(key) != 32 {
		return nil, ErrInvalidKeyLength{Length: len(key)}
	}
	return &CredentialCrypto{key: key}, nil
}

// NewFromBase64Key parses the RELAY_MASTER_KEY env var: base64 for exactly
// 32 raw bytes (AES-256) — identical parsing to
// CredentialCrypto.init(base64Key:).
func NewFromBase64Key(base64Key string) (*CredentialCrypto, error) {
	data, err := base64.StdEncoding.DecodeString(base64Key)
	if err != nil {
		return nil, ErrInvalidBase64
	}
	return New(data)
}

// GenerateBase64Key generates a fresh random 32-byte key, base64-encoded —
// what docs/relay-deployment.md tells operators to run once (`openssl rand
// -base64 32`-equivalent) to produce RELAY_MASTER_KEY.
func GenerateBase64Key() (string, error) {
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(key), nil
}

func (c *CredentialCrypto) aead() (cipher.AEAD, error) {
	block, err := aes.NewCipher(c.key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

// Encrypt seals plaintext, returning nonce||ciphertext||tag as a single
// []byte — the same "combined" layout CredentialCrypto.encrypt(_:)
// produces, and the only thing the store needs to persist per credential.
func (c *CredentialCrypto) Encrypt(plaintext string) ([]byte, error) {
	aead, err := c.aead()
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, nonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	// Seal(dst, nonce, plaintext, additionalData): appends ciphertext||tag
	// to dst. Passing nonce as dst yields nonce||ciphertext||tag in one
	// slice, matching swift-crypto's AES.GCM.SealedBox.combined exactly.
	combined := aead.Seal(nonce, nonce, []byte(plaintext), nil)
	return combined, nil
}

// Decrypt reverses Encrypt (and, critically, any AES.GCM.seal-produced
// combined representation from the retired Swift relay). Returns ErrDecryptionFailed
// on any authentication-tag mismatch, malformed input, or non-UTF-8
// plaintext.
func (c *CredentialCrypto) Decrypt(combined []byte) (string, error) {
	if len(combined) < nonceSize+tagSize {
		return "", ErrDecryptionFailed
	}
	aead, err := c.aead()
	if err != nil {
		return "", ErrDecryptionFailed
	}
	nonce := combined[:nonceSize]
	ciphertextAndTag := combined[nonceSize:]
	plaintext, err := aead.Open(nil, nonce, ciphertextAndTag, nil)
	if err != nil {
		return "", ErrDecryptionFailed
	}
	// Swift's CredentialCrypto.decrypt fails when the plaintext isn't
	// valid UTF-8 (String(data:encoding:) returning nil) — Go strings can
	// carry arbitrary bytes, so check explicitly for identical behavior.
	if !utf8.Valid(plaintext) {
		return "", ErrDecryptionFailed
	}
	return string(plaintext), nil
}
